pkg/sfu/redreceiver.go

```diff
@@ -118,17 +119,26 @@ func (r *RedReceiver) encodeRedForPrimary(pkt *rtp.Packet, redPayload []byte) (i
 
 	for _, prev := range r.pktBuff[lastNilPkt+1:] {
 		if pkt.SequenceNumber == prev.SequenceNumber ||
-			(pkt.SequenceNumber-prev.SequenceNumber) > uint16(maxRedCount) {
+			(pkt.SequenceNumber-prev.SequenceNumber) > uint16(redLength) ||
+			(pkt.Timestamp-prev.Timestamp) >= (1<<14) {
 			continue
 		}
 		redPkts = append(redPkts, prev)
 	}
 
-	if r.pktBuff[1] == nil || pkt.SequenceNumber-r.pktBuff[1].SequenceNumber < 0x8000 {
-		/* update packet, not copy the rtp packet here since we only hold two packets for red encoding,
-		the upstream buffer size is much larger than two, so it is safe to use packet directly
-		*/
-		r.pktBuff[0], r.pktBuff[1] = r.pktBuff[1], pkt
+	// insert primary packet in history buffer
+	// NOTE: packet is copied from retransmission buffer and used in forwarding path. So, not making another
+	// copy here and just maintaining pointer to the packet as the forwarding path should not alter the packet.
+	for i := redLength - 1; i >= 0; i-- {
+		if r.pktBuff[i] == nil || // history is empty
+			pkt.SequenceNumber-r.pktBuff[i].SequenceNumber < (1<<15) { // received packet has more recent sequence number
+			// age out older ones
+			for j := 0; j < i; j++ {
+				r.pktBuff[j] = r.pktBuff[j+1]
+			}
+			r.pktBuff[i] = pkt
+			break
+		}
```

Comment:
Re-doing this a bit to catch scenarios like following
- Packet arrival sequence is 5, 7, 6
- When 6 is received, the history would have 5, 7
- RED encoding would have been 5, 6 (only one redundancy)
- Previously, the history would not have been changed as 6 is out-of-order with the last element. So, when 8 was received, it would have been encoded as 7, 8 (only one redundancy as 5 is too old)

Changing it to walk back and find a location for the new packet. So, after 6, the history would be 6, 7 and when 8 comes around it can be encoded with 6, 7, 8 (two redundancies)